rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -20,7 +20,41 @@ use hyperlane_core::HyperlaneDomain;
 
 use super::pending_operation::*;
 
-type OpQueue = Arc<Mutex<BinaryHeap<Reverse<Box<DynPendingOperation>>>>>;
+/// Queue of generic operations that can be submitted to a destination chain.
+/// Includes logic for maintaining queue metrics by the destination and `app_context` of an operation
+#[derive(Debug, Clone, new)]
+struct OpQueue {
+    metrics: IntGaugeVec,
+    queue_metrics_label: String,
+    #[new(default)]
+    queue: Arc<Mutex<BinaryHeap<Reverse<Box<DynPendingOperation>>>>>,
+}
+
+impl OpQueue {
+    /// Push an element onto the queue and update metrics
+    async fn push(&self, op: Box<DynPendingOperation>) {
+        // increment the metric before pushing onto the queue, because we lose ownership afterwards
+        self.get_operation_metric(&op).inc();
+
+        self.queue.lock().await.push(Reverse(op));
+    }
+
+    /// Pop an element from the queue and update metrics
+    async fn pop(&self) -> Option<Reverse<Box<DynPendingOperation>>> {
+        let op = self.queue.lock().await.pop();
+        op.map(|op| {
+            self.get_operation_metric(&op.0).dec();
```

Comment:
one pretty big downside of my suggestion to update on the pop/push "entrypoints" is that this would update the metric as if the operation that's currently being processed isn't part of the queue

wondering if this can get confusing if e.g. we have one message "stuck" in the queue that we're concerned about, but each time we try to process it the metric shows the queue size as 0, so when looking at alerts / dashboards it looks like we're successfully processing the entire queue

wdyt? i guess we could keep as is, or consider just updating the metrics upon `PendingOperationResult::Success` or something

Reply:
there's usually several messages that fail to deliver at once, and the queue size would still increase in those cases - the queue may just be off by one sometimes. In those cases the sum of `prepare_queue` and `confirm_queue` will still be nonzero I believe. I think that's good enough, added a comment about this

Reply:
fwiw during the period that a message is being "worked on" (like building metadata, or waiting for a tx receipt, etc), the message won't be a part of the confirm queue yet, so the sum will be 0. You're right that usually there's more than one in the queue at least at our scale now though. So happy to merge and we can always keep an eye if it leads to confusion